monai/transforms/compose.py

```diff
@@ -41,7 +41,84 @@
 
 logger = get_logger(__name__)
 
-__all__ = ["Compose", "OneOf", "RandomOrder", "SomeOf", "execute_compose"]
+__all__ = [
+    "Compose",
+    "OneOf",
+    "RandomOrder",
+    "SomeOf",
+    "compose_iterator",
+    "ranged_compose_iterator",
+    "execute_compose",
+]
+
+
+def compose_iterator(compose, step_into_all=False):
+    """
+    ``compose_iterator`` is a function that returns an iterator over the transforms in a ``Compose``
+    instance.
+
+    It works by iterating over transforms, and stepping into nested ``Compose`` instances in a
+    recursive fashion. Given the following example:
+
+    ..code-block:: python
+
+    ::
+        t1, t2, t3, t4 = # some transforms
+        c = Compose([t1, Compose([t2, t3]), t4])
+        print([tx for tx in compose_iterator(c)])
+
+        > [t1, t2, t3, t4]
+
+    By default, ``compose_iterator`` doesn't step into ``OneOf``, ``SomeOf``, ``RandomOrder`` or
+    other classes that inherit ``Compose``. If ``step_into_all`` is passed True, it will step into
+    any Compose-like object as if it were just a container. As such, if you are using ``compose_iterator``
+    to step over objects as if executing them, ``step_into_all`` should be False. If you are using
+    ``compose_iterator`` to step over the static structure of the pipeline (such as if you are setting
+    or clearing flags on every transform), you should set ``step_into_all`` to True
+
+    Args:
+        compose: A `Compose` instance that contains the transforms to be iterated over
+        step_into_all: A boolean flag that indicates whether to step into randomised `Compose`
+            sub-classes such as `OneOf`, `SomeOf`, and `RandomOrder`
+    """
+
+    if not isinstance(compose, Compose):
+        raise ValueError(f"must be of type {(type.Compose)} but is type {type(compose)}")
+
+    for i in range(len(compose.transforms)):
+        tx = compose.transforms[i]
+        if type(tx) == Compose or (step_into_all is True and isinstance(tx, Compose)):
+            yield from compose_iterator(tx)
+        else:
+            yield tx
+
+
+def ranged_compose_iterator(compose, start=None, end=None, step_into_all=False):
```

Comment:
Would [islice](https://docs.python.org/3/library/itertools.html#itertools.islice) do the same?

Reply:
No, because we have the special semantic that there are containers that we don't step into.